tests/unit/test_database.py

```diff
@@ -0,0 +1,115 @@
+from ops.testing import Harness
+
+from charm import MySQLOperatorCharm
+from constants import DB_RELATION_NAME
+from tests.unit.helpers import patch_network_get
+
+
+class TestDatase(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(MySQLOperatorCharm)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.begin()
+        self.peer_relation_id = self.harness.add_relation("database-peers", "database-peers")
+        self.harness.add_relation_unit(self.peer_relation_id, "mysql/1")
+        self.database_relation_id = self.harness.add_relation(DB_RELATION_NAME, "app")
+        self.harness.add_relation_unit(self.database_relation_id, "app/0")
+        self.charm = self.harness.charm
+
+    @patch_network_get(private_address="1.1.1.1")
+    @patch("mysqlsh_helpers.MySQL.get_mysql_version", return_value="8.0.29-0ubuntu0.20.04.3")
+    @patch(
+        "mysqlsh_helpers.MySQL.get_cluster_members_addresses",
+        return_value={"2.2.2.1:3306", "2.2.2.3:3306", "2.2.2.2:3306"},
+    )
+    @patch("mysqlsh_helpers.MySQL.get_cluster_primary_address", return_value="2.2.2.2:3306")
+    @patch("mysqlsh_helpers.MySQL.create_application_database_and_scoped_user")
+    @patch("relations.database.generate_random_password", return_value="super_secure_password")
+    def test_database_requested(
+        self,
+        _generate_random_password,
+        _create_application_database_and_scoped_user,
+        _get_cluster_primary_address,
+        _get_cluster_members_addresses,
+        _get_mysql_version,
+    ):
+        # run start-up events to enable usage of the helper class
+        self.harness.set_leader(True)
+        self.charm.on.config_changed.emit()
+
+        # confirm that the relation databag is empty
+        database_relation_databag = self.harness.get_relation_data(
+            self.database_relation_id, self.harness.charm.app
+        )
+        database_relation = self.charm.model.get_relation(DB_RELATION_NAME)
+        app_unit = list(database_relation.units)[0]
+
+        # simulate cluster initialized by editing the flag
+        self.harness.update_relation_data(
+            self.peer_relation_id, self.charm.app.name, {"units-added-to-cluster": "1"}
+        )
+
+        self.assertEqual(database_relation_databag, {})
+        self.assertEqual(database_relation.data.get(app_unit), {})
+        self.assertEqual(database_relation.data.get(self.charm.unit), {})
+
+        # update the app leader unit data to trigger database_requested event
+        self.harness.update_relation_data(
+            self.database_relation_id, "app", {"database": "test_db"}
+        )
+
+        # confirm that the relation databag is updated
+        database_relation_databag = self.harness.get_relation_data(
+            self.database_relation_id, self.harness.charm.app
+        )
+        database_relation = self.charm.model.get_relation(DB_RELATION_NAME)
+        app_unit = list(database_relation.units)[0]
```

Comment:
Is this code block duplicated? (L51-L52)

Reply:
I was checking if people were paying attention on reviews ;)